libp2p/src/builder.rs

```diff
@@ -0,0 +1,271 @@
+        Error = Self::E,
+        Dial = Self::D,
+        ListenerUpgrade = Self::U,
+        Output = (libp2p_identity::PeerId, StreamMuxerBox),
+    > + Send
+    + Unpin
+    + 'static
+{
+    type E: Send + Sync + 'static;
+    type D: Send;
+    type U: Send;
+}
+
+impl<T> BuildableTransport for T
+where
+    T: Transport<Output = (libp2p_identity::PeerId, StreamMuxerBox)> + Send + Unpin + 'static,
+    <T as Transport>::Error: Send + Sync + 'static,
+    <T as Transport>::Dial: Send,
+    <T as Transport>::ListenerUpgrade: Send,
+{
+    type E = T::Error;
+    type D = T::Dial;
+    type U = T::ListenerUpgrade;
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    #[cfg(all(feature = "tokio", feature = "tcp"))]
+    fn tcp() {
+        let key = libp2p_identity::Keypair::generate_ed25519();
+        let _: libp2p_core::transport::Boxed<(libp2p_identity::PeerId, StreamMuxerBox)> =
+            TransportBuilder::new(key)
+                .with_tokio()
+                .with_tcp()
+                .without_relay()
+                .without_dns()
+                .build();
+    }
+
+    #[test]
+    #[cfg(all(feature = "tokio", feature = "tcp", feature = "relay"))]
+    fn tcp_relay() {
+        let key = libp2p_identity::Keypair::generate_ed25519();
+        let (relay_transport, _) = libp2p_relay::client::new(key.public().to_peer_id());
+
+        let _: libp2p_core::transport::Boxed<(libp2p_identity::PeerId, StreamMuxerBox)> =
+            TransportBuilder::new(key)
+                .with_tokio()
+                .with_tcp()
+                .with_relay(relay_transport)
+                .without_dns()
+                .build();
+    }
+
+    #[test]
+    #[cfg(all(feature = "tokio", feature = "tcp", feature = "dns"))]
+    fn tcp_dns() {
+        let key = libp2p_identity::Keypair::generate_ed25519();
+        let _: libp2p_core::transport::Boxed<(libp2p_identity::PeerId, StreamMuxerBox)> =
+            TransportBuilder::new(key)
+                .with_tokio()
+                .with_tcp()
+                .without_relay()
+                .with_dns()
+                .build();
+    }
+
+    /// Showcases how to provide custom transports unknown to the libp2p crate, e.g. QUIC or WebRTC.
+    #[test]
+    #[cfg(all(feature = "tokio", feature = "tcp"))]
+    fn tcp_other_transport_other_transport() {
+        let key = libp2p_identity::Keypair::generate_ed25519();
+        let _: libp2p_core::transport::Boxed<(libp2p_identity::PeerId, StreamMuxerBox)> =
+            TransportBuilder::new(key)
+                .with_tokio()
+                .with_tcp()
+                .without_relay()
+                .without_dns()
+                .with_other_transport(libp2p_core::transport::dummy::DummyTransport::new())
+                .with_other_transport(libp2p_core::transport::dummy::DummyTransport::new())
+                .with_other_transport(libp2p_core::transport::dummy::DummyTransport::new())
+                .build();
+    }
```

Comment:
`TransportBuilder` allows users to bring their own `Transport` implementations, e.g. `libp2p-quic`, even though they are unknown to `libp2p`. The only requirement is that they implement `Transport<Output = (PeerId, StreamMuxerBox)>`.